Introduce new Generic Fragments, ViewModels, and Adapters for the display of API data.

tzugen opened the pull request:
* Splits former SelectAlbumFragment into separate fragments for Albums and general collections of tracks
* Renames and refactors SelectArtist view to extend the new Generic classes
* Adds error handling (Fixes #484)
* Adds EndlessScrolling capabilities to all Album Lists :partying_face: 
* Uses RecyclerViews and LiveData for performance and in-memory caching
* Refreshes the UI to be aligned with the ArtistList UI
* Add a new GenericEntry to the domain data classes, and make other types extend it

tzugen committed:
Rename classes in preparation of split.

tzugen committed:
Introduce new Generic Fragments, ViewModels, and Adapters for the display of API data.

* Splits former SelectAlbumFragment into separate fragments for Albums and general collections of tracks
* Renames and refactors SelectArtist view to extend the new Generic classes
* Adds error handling (Fixes #484)
* Adds EndlessScrolling capabilities to all Album Lists
* Uses RecyclerViews and LiveData for performance and in-memory caching
* Refreshes the UI to be aligned with the ArtistList UI
* Add a new GenericEntry to the domain data classes, and make other types extend it

tzugen committed:
Use default values for LongMethod, LongParameterList, LargeClass and ComplexMethod.
The algrorithms in detekt had changed when I updated detekt to v1.16, but our values not,
resulting in a very long baseline file with many non-sensical warnings.
By removing our config we follow the detekt config when they modify their algorithms again.

tzugen committed:
Regenerate baseline files

nitehu commented on ultrasonic/src/main/kotlin/org/moire/ultrasonic/fragment/AlbumRowAdapter.kt:
```diff
@@ -0,0 +1,100 @@
+import org.moire.ultrasonic.view.SelectMusicFolderView
+
+/**
+ * Creates a Row in a RecyclerView which contains the details of an Artist
```
I think "Artist" here is a copy/paste error (and in the copyright header too)

nitehu commented on ultrasonic/src/main/kotlin/org/moire/ultrasonic/fragment/AlbumListModel.kt:
```diff
@@ -0,0 +1,94 @@
+        return !isOffline() && !Util.getShouldUseId3Tags() && isAlphabetical
+    }
+
+    private fun sortableCollection(albumListType: String): Boolean {
```
The name of this confused me for a second, maybe isCollectionSortable would be better?

nitehu commented on ultrasonic/src/main/kotlin/org/moire/ultrasonic/fragment/TrackCollectionFragment.kt:
```diff
@@ -229,6 +231,14 @@ class SelectAlbumFragment : Fragment() {
     }
 
+    val handler = CoroutineExceptionHandler { _, exception ->
+        println("CoroutineExceptionHandler got $exception")
```
I think this println was supposed to be a Timber log write...

nitehu commented on ultrasonic/src/main/kotlin/org/moire/ultrasonic/fragment/ArtistListFragment.kt:
```diff
@@ -0,0 +1,72 @@
+
+    override val musicFolderObserver = { changedFolders: List<MusicFolder> ->
+        viewAdapter.notifyDataSetChanged()
+        selectFolderHeader!!.setData(
```
This can be null, when there is no header!
Currently it can throw an exception:

```
E/AndroidRuntime: FATAL EXCEPTION: main
    Process: org.moire.ultrasonic.debug, PID: 28954
    java.lang.NullPointerException
        at org.moire.ultrasonic.fragment.ArtistListFragment$musicFolderObserver$1.invoke(ArtistListFragment.kt:67)
        at org.moire.ultrasonic.fragment.ArtistListFragment$musicFolderObserver$1.invoke(ArtistListFragment.kt:16)
        at org.moire.ultrasonic.fragment.GenericListFragment$sam$androidx_lifecycle_Observer$0.onChanged(Unknown Source:6)
        at androidx.lifecycle.LiveData.considerNotify(LiveData.java:131)
        at androidx.lifecycle.LiveData.dispatchingValue(LiveData.java:149)
        at androidx.lifecycle.LiveData.setValue(LiveData.java:307)
        at androidx.lifecycle.MutableLiveData.setValue(MutableLiveData.java:50)
        at androidx.lifecycle.LiveData$1.run(LiveData.java:91)
        at android.os.Handler.handleCallback(Handler.java:938)
        at android.os.Handler.dispatchMessage(Handler.java:99)
        at android.os.Looper.loop(Looper.java:223)
        at android.app.ActivityThread.main(ActivityThread.java:7656)
        at java.lang.reflect.Method.invoke(Native Method)
        at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:592)
        at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:947)
```

nitehu commented on ultrasonic/src/main/kotlin/org/moire/ultrasonic/fragment/AlbumListFragment.kt:
```diff
@@ -0,0 +1,110 @@
+    /**
+     * Whether to show the folder selector
+     */
+    override var folderHeaderEnabled = false
```
I think this shouldn't be always false here, at least now the folder selector for the albums doesn't work at all...

nitehu reviewed (changes requested):
Great job! The endless scrolling works very good, and I didn't find any noticable problems with the layout.
There are some minor issues, mostly with the folder selector. Could you please check these?

tzugen committed:
Correctly display the folder header also in album view

Also fix a number of smaller issues

tzugen committed:
Create a new SelectMusicFolderViewHolder instance when the Adapter is recreated,
instead of reusing an old one. This lead to a crash.

Also make the AlbumList use the correct placeholder,
remove an unused resource file, and update the placeholder to look less clumsy.
